File: app/src/main/java/com/ricamgar/notify/addreminder/AddReminderPresenter.java
package com.ricamgar.notify.addreminder;

import com.ricamgar.notify.base.mvp.AbstractBasePresenter;
import com.ricamgar.notify.domain.reminder.model.Reminder;
import com.ricamgar.notify.domain.reminder.usecase.AddOrUpdateReminder;

import javax.inject.Inject;

public class AddReminderPresenter extends AbstractBasePresenter<Reminder> {

    private final AddOrUpdateReminder addReminderUseCase;
    private final Reminder.Builder reminderBuilder = new Reminder.Builder();

    @Inject
    public AddReminderPresenter(AddOrUpdateReminder addReminderUseCase) {
        this.addReminderUseCase = addReminderUseCase;
    }

    public void insertReminder() {
        Reminder reminder = reminderBuilder.build();
        if (reminder.description.isEmpty()) {
            ((View) view).showError("Invalid reminder");
            return;
        }

        addSubscription(addReminderUseCase.setReminder(reminder)
                .execute()
                .subscribe(
                        insertedReminder -> {
                            ((View) view).showReminderCreatedSuccess();
                            ((View) view).closeDialog();
                        },
                        throwable -> ((View) view).showError(throwable.getMessage())
                ));
    }

    public void setDescription(CharSequence description) {
        reminderBuilder.description(description.toString());
    }

    public interface View extends BaseView {

        void closeDialog();

        void showError(String error);

        void showReminderCreatedSuccess();
    }
}

File: app/src/main/java/com/ricamgar/notify/addreminder/AddReminderPresenter.kt
package com.ricamgar.notify.addreminder

import com.ricamgar.notify.base.mvp.AbstractBasePresenter
import com.ricamgar.notify.domain.reminder.model.Reminder
import com.ricamgar.notify.domain.reminder.usecase.AddOrUpdateReminder
import io.reactivex.Scheduler

import javax.inject.Inject
import javax.inject.Named

class AddReminderPresenter
@Inject constructor(
  private val addReminderUseCase: AddOrUpdateReminder,
  @Named("io") ioThread: Scheduler,
  @Named("main") mainThread: Scheduler
) : AbstractBasePresenter<Reminder>(ioThread, mainThread) {

  private lateinit var description: String

  fun insertReminder() {
    if (description.isEmpty()) {
      (view as View).showError("Invalid reminder")
      return
    }

    val reminder = Reminder(0, description, false, 0)
    addDisposable(addReminderUseCase.execute(reminder)
      .compose(singleSchedulers())
      .subscribe({
        (view as View).showReminderCreatedSuccess()
        (view as View).closeDialog()
      }, { throwable -> (view as View).showError(throwable.message!!) }
      ))
  }

  fun setDescription(description: CharSequence) {
    this.description = description.toString()
  }

  interface View : AbstractBasePresenter.BaseView {

    fun closeDialog()

    fun showError(error: String)

    fun showReminderCreatedSuccess()
  }
}
